src/main/java/slimeknights/tconstruct/smeltery/tileentity/CastingTileEntity.java

```diff
@@ -0,0 +1,236 @@
+      // can have ItemStacks with stacksize > 1 as output
+      // we therefore spill the whole contents on extraction.
+      ItemStack stack = getStackInSlot(slot);
+      if (slot == OUTPUT) {
+        // fire player smelt event?
+      }
+      ItemHandlerHelper.giveItemToPlayer(player, stack);
+      setInventorySlotContents(slot, ItemStack.EMPTY);
+
+      // send a block update for the comparator, needs to be done after the stack is removed
+      if (slot == OUTPUT) {
+        this.getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+      }
+    }
+  }
+
+
+  @Override
+  @Nonnull
+  public int[] getSlotsForFace(Direction side) {
+    return new int[]{INPUT, OUTPUT};
+  }
+
+  @Override
+  public boolean canInsertItem(int index, ItemStack itemStackIn, @Nullable Direction direction) {
+    return index == INPUT && !isStackInSlot(OUTPUT);
+  }
+
+  @Override
+  public boolean canExtractItem(int index, ItemStack stack, Direction direction) {
+    return index == 1;
+  }
+
+  @Override
+  public void tick() {
+    // no recipe
+    if (recipe == null) {
+      return;
+    }
+    // fully filled
+    if (tank.getFluidAmount() == tank.getCapacity() && !tank.getFluid().isEmpty()) {
+      timer++;
+      if (!getWorld().isRemote) {
+        if (timer >= recipe.getCoolingTime()) {
+          if (recipe.consumesCast()) {
+            setInventorySlotContents(INPUT, ItemStack.EMPTY);
+          }
+          if (recipe.switchSlots()) {
+            setInventorySlotContents(INPUT, recipe.getRecipeOutput());
+          }
+          else {
+            setInventorySlotContents(OUTPUT, recipe.getRecipeOutput());
+          }
+          getWorld().playSound(null, pos, SoundEvents.BLOCK_LAVA_EXTINGUISH, SoundCategory.AMBIENT, 0.07f, 4f);
+
+          reset();
+
+          getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+        }
+      }
+      else if (getWorld().rand.nextFloat() > 0.9f) {
+        world.addParticle(ParticleTypes.SMOKE, pos.getX() + getWorld().rand.nextDouble(), pos.getY() + 1.1d, pos.getZ() + getWorld().rand.nextDouble(), 0.0D, 0.0D, 0.0D);
+      }
+    }
+  }
+
+  protected AbstractCastingRecipe findRecipe(Fluid fluid) {
+
+    recipes.clear();
+    recipes = world.getRecipeManager().getRecipes((IRecipeType<AbstractCastingRecipe>) recipeType, this, world);
+
+    for (AbstractCastingRecipe candidate : recipes) {
+      if (candidate.matches(fluid, this, world)) {
+        recipe = candidate;
+      }
+    }
+//    CastingRecipe recipe = getWorld().getRecipeManager().getRecipe(TinkerSmeltery.castingRecipeType, this, this.world).orElse(null);
+    return recipe;
+  }
+
+  public int initNewCasting(Fluid fluid, IFluidHandler.FluidAction action) {
+    AbstractCastingRecipe recipe = findRecipe(fluid);
+    if (recipe != null) {
+      if (action == IFluidHandler.FluidAction.SIMULATE) {
```

Comment:
Simulate means don't change anything

Reply:
Let me expand upon this one: if the `recipe` variable is used in game logic, such as the tick method, it should never be changed during simulate. Anyone can simulate and the block should not be affected, one common example is for interfaces that read fluid data in a tank. If its a "last recipe" cache only used in this method, that is fine, but then the variable should not be referenced in `tick`.

As I mentioned on Discord, best solution is two recipes, `currentRecipe` is the one that `tick` and `reset` will handle, you set it when we get an `EXECUTE`. It can block logic as needed and all that fun. `lastRecipe` is our simple cache so if we are trying to craft the same thing as last time, we save time in lookup. `lastRecipe` will only be modified here, and is modified regardless of the action. `currentRecipe` is modified in multiple places.